File: 0130-surrounded-regions/0130-surrounded-regions.cpp
class Solution {
public:
    void dfs(vector<vector<char>>& board, int m, int n, int i, int j){
        //edge case 
        if(i < 0 || j < 0 || i >= n || j >= m || board[i][j] != 'O')
        {
            return;
        }

        board[i][j] = '#';

        dfs(board,m,n,i-1,j); //up
        dfs(board,m,n,i+1,j); // down
        dfs(board,m,n,i,j-1); // <- west
        dfs(board,m,n,i,j+1); // east->

    }
    void solve(vector<vector<char>>& board) {
        int r = board.size();
        int c = board[0].size();

        //Moving to first and last column
        for(int i = 0; i < r; i++){
            if(board[i][0] == 'O'){
                dfs(board,c,r,i,0);
            }
            if(board[i][c-1] == 'O'){
                dfs(board,c,r,i,c-1);
            }
        }

        //Moving to first and last row
        for(int j = 0; j < c; j++){
            if(board[0][j] == 'O'){
                dfs(board,c,r,0,j);
            }
            if(board[r-1][j] == 'O'){
                dfs(board,c,r,r-1,j);
            }
        }

        //Now as our boundry O are marked with #, so if we find any # replace it with O and for rest of the case we can replace it with X
        for(int i = 0; i<r; i++){
            for(int j = 0; j < c; j++){
                if(board[i][j] == 'O'){
                    board[i][j] = 'X';
                }
                if(board[i][j] == '#'){
                    board[i][j] = 'O';
                }
            }
        }
    }
};

File: 0130-surrounded-regions/0130-surrounded-regions.java
class Solution {
    void dfs(char[][] board, int n, int m, int currRow, int currCol){
        if(currRow < 0 || currRow >= n || currCol < 0 || currCol >= m || board[currRow][currCol] != 'O'){
            return ;
        }

        board[currRow][currCol] = '#';
        dfs(board, n, m, currRow-1, currCol); // up
        dfs(board, n, m, currRow+1, currCol); // down
        dfs(board, n, m, currRow, currCol+1); // <- right
        dfs(board, n, m, currRow, currCol-1); // left ->
        
    }


    public void solve(char[][] board) {
        int n = board.length;
        int m = board[0].length;

        for(int r = 0; r < n; r++){
            if(board[r][0] == 'O'){
                dfs(board, n, m, r, 0);
            }

            if(board[r][m-1] == 'O'){
                dfs(board, n, m, r, m-1);
            }
        }


        for(int c = 0; c < m; c++){
            if(board[0][c] == 'O'){
                dfs(board, n, m, 0, c);
            }

            if(board[n-1][c] == 'O'){
                dfs(board, n, m, n-1, c);
            }
        }

        for(int r = 0; r < n; r++){
            for(int c = 0; c < m; c++){
                if(board[r][c] == 'O'){
                    board[r][c] = 'X';
                }

                if(board[r][c] == '#'){
                    board[r][c] = 'O';
                }
            }
        }
        
    }
}